### PR TITLE
Add AQL driver

### PR DESCRIPTION
**Pull request by mrshankly**
In the context of my master thesis I implemented an AQL driver that I'd like to contribute to the upstream repository. The implementation is based on the cassandra driver. Let me know if there's anything I need to change or improve.

Thanks

### CODE REVIEW

File: priv/build_schema.aql
Comment (goncalotomas) [inline priv/build_schema.aql:1]: I know this wasn't done for the existing CQL file, but maybe it would make sense to have a 1-2 lines explaining what is AQL, maybe linking to the source or a documentation page. It would help distinguish what essentially are 2 almost identical files 🙂

File: rebar.config
Comment (goncalotomas) [inline rebar.config:15]: Are you planning on tagging a particular version of the AQL Client or publishing it on hex.pm?  I would rather reference a stable release than the master branch.

Periodically I try to update dependencies to their latest version, and typically I try out different patch and minor versions. If for some reason you wish to add breaking changes, this will break FMKe's version of aqlc down the line. Another thing this helps is to lock down the versions of the dependencies of `aqlc`.

If you need some help for this I'd be happy to walk you through the release process 😄

File: rebar.lock
Comment (goncalotomas) [inline rebar.lock:20]: The TravisCI build seems to be failing on building this dependency. Any ideas why? I didn't have time to look into it yet, but this is something that needs to be fixed before merging, whether it's the build environment or something in `paillier`.

File: src/fmke_driver_opt_aql.erl
Comment (goncalotomas) [inline src/fmke_driver_opt_aql.erl:2]: Feel free to add a link to any documentation about AQL here as well, if applicable.

File: rebar.lock
Comment (goncalotomas) [inline rebar.lock:44]: Was this entry added automatically by rebar3? What is this entry for? It mentions other versions of already existing dependencies (e.g. riak_pb). Why?

**Review by goncalotomas — CHANGES_REQUESTED** [review]
Overall, great work! There are just some operational concerns related to releases and versions that I want to get resolved to guarantee that this remains supported for a long time.

Aside from my comments, do you see any need for changing `README.md` and updating the supported databases? AQL is an add-on for AntidoteDB, but maybe it would be worth noting that AQL is supported for AntidoteDB. What do you think?

**Comment by mrshankly** [comment]
> Aside from my comments, do you see any need for changing `README.md` and updating the supported databases? AQL is an add-on for AntidoteDB, but maybe it would be worth noting that AQL is supported for AntidoteDB. What do you think?

Sorry for the delay in getting back to this. Yes I think it's a good idea, I've added a note to `README.md` saying that the AQL interface is also supported.

File: priv/build_schema.aql
Comment (mrshankly) [reply priv/build_schema.aql:9]: Done, I've added a link to the most recent AQL repository.

File: rebar.lock
Comment (mrshankly) [reply rebar.lock:40]: Yes, all modifications to this file were automatically added by rebar3. I think it's because I'm using a more recent version of rebar3, or at least that's what some warnings in the travis build log suggest. Should I use a specific version of rebar3?

File: rebar.lock
Comment (mrshankly) [reply rebar.lock]: This was my mistake, it's fixed now.

File: src/fmke_driver_opt_aql.erl
Comment (mrshankly) [reply src/fmke_driver_opt_aql.erl:2]: Done.

File: rebar.config
Comment (mrshankly) [reply rebar.config]: I've published it on hex.pm, I've updated the `rebar.config` file to use that package.

**Comment by mrshankly** [comment]
Travis seems to fail when running the common test suites, not really sure why...

**Comment by goncalotomas** [comment]
I'm going to try to look into it this weekend. It's possible that the travis config needs an update.

**Comment by bieniusa** [comment]
Might be worth to move from Travis to Github Workflow - seems to work well and is better integrated in the Github ecosystem.

**Comment by goncalotomas** [comment]
I've began the process of moving from TravisCI to GitHub actions and I'm seeing the same type of errors in the antidote suites, where it is triggering the timetrap (time limit for tests to be run). This is because the Antidote client tries to connect over and over without a timeout specified, meaning that if the connection fails for some reason (antidote not started, wrong port, etc), it will not return any feedback...

Seeing as this is a temporary problem with Antidote's client library, I am willing to merge these changes as long as they have been properly tested locally. @mrshankly did you test these locally? Are you willing to share some test runs you made successfully?
I also just noticed that in `make coverage` there isn't a scenario run for AQL, maybe we should add one.

**Comment by mrshankly** [comment]
> Seeing as this is a temporary problem with Antidote's client library, I am willing to merge these changes as long as they have been properly tested locally. @mrshankly did you test these locally? Are you willing to share some test runs you made successfully?

I did, the benchmark runs, and without errors. However, the results aren't very good, but that's because of AQL I believe. Some operations take a very long time to complete because of the lack of indexes with composite keys, and as more data is inserted into the database, these operations get even slower. Here is the summary produced by a benchmark run:

summary

> I also just noticed that in `make coverage` there isn't a scenario run for AQL, maybe we should add one.

Yeah, that's a good idea. I'll try to add one in the next days.

**Comment by mrshankly** [comment]
Hi @goncalotomas, just a small update.

I haven't forgotten about this pull request, I'm hoping to have everything ready soon. I've been busy lately and just haven't got the time to work on this. Sorry for the delay.

**Comment by mrshankly** [comment]
Added a ct scenario for AQL. I followed the implementation of the existing tests and added an extra make rule, `make ct-aql`. I created a repository on docker hub with the AQL image, [here](https://hub.docker.com/r/jbmarques/aql).

Let me know if there's anything else left to do.
Thanks.

**Comment by goncalotomas** [comment]
> I can squash the commits into one if you want.
Please do, if you could do it that'd be great!

This weekend I'll try to fix the coverage and pipeline situation before merging this, but if I can't I'll just merge it first and deal with CI/CD + coverage later. Looks like a solid MR!

**Comment by mrshankly** [comment]
When I implementation this driver I used the cassandra driver as a reference, but I noticed that some operations could be implemented a little more efficiently in the AQL driver. Specifically, there are a few read operations here and there, that aren't really necessary. I'll update those today.

**Comment by mrshankly** [comment]
I've removed the read operations that were unnecessary.

I've also added an alternative build schema script, `priv/build_schema_fk.aql`, that uses the referential integrity mechanism of AQL if one wishes to benchmark AQL with that. There is a note in the README that explains this.

Feel free to merge this when you're ready.
Thanks.

**Review by goncalotomas — APPROVED** [review]

**Comment by goncalotomas** [comment]
CI was already broken, is to be fixed on another MR.
João, this is great work and thanks for sticking with me for the review changes I requested. Glad to be merging this into FMKe.

Thanks for your commitment!

**Pull request merged**
